src/lib/operators/window_function_evaluator.cpp

```diff
@@ -0,0 +1,607 @@
+    using State = typename Impl::State;
+    auto state = State{};
+
+    const RelevantRowInformation* previous_row = nullptr;
+
+    for (const auto& row : hash_partition) {
+      if (previous_row && std::is_eq(compare_with_null_equal(previous_row->partition_values, row.partition_values))) {
+        Impl::update_state(state, row);
+      } else {
+        state = Impl::initial_state(row);
+      }
+      emit_computed_value(row.row_id, Impl::current_value(state));
+      previous_row = &row;
+    }
+  });
+}
+
+namespace {
+
+// Sentinel type to indicate that Range mode is used without an order-by column.
+struct NoOrderByColumn {};
+
+template <FrameType frame_type, typename OrderByColumnType>
+struct WindowBoundCalculator {
+  static QueryRange calculate_window_bounds([[maybe_unused]] std::span<const RelevantRowInformation> partition,
+                                            [[maybe_unused]] uint64_t tuple_index,
+                                            [[maybe_unused]] const FrameDescription& frame) {
+    std::string error_message = "Unsupported frame type and order-by column type combination: ";
+    error_message += [&]() {
+      using std::literals::string_view_literals::operator""sv;
+
+      switch (frame_type) {
+        case FrameType::Rows:
+          return "Rows"sv;
+        case FrameType::Range:
+          return "Range"sv;
+        case FrameType::Groups:
+          return "Groups"sv;
+        default:
+          return ""sv;
+      }
+    }();
+    error_message += " ";
+    error_message += data_type_to_string.left.at(data_type_from_type<OrderByColumnType>());
+    Fail(error_message);
+  }
+};
+
+template <typename OrderByColumnType>
+struct WindowBoundCalculator<FrameType::Rows, OrderByColumnType> {
+  static QueryRange calculate_window_bounds(std::span<const RelevantRowInformation> partition, uint64_t tuple_index,
+                                            const FrameDescription& frame) {
+    const auto start = frame.start.unbounded ? 0 : clamped_sub<uint64_t>(tuple_index, frame.start.offset, 0);
+    const auto end = frame.end.unbounded ? partition.size()
+                                         : clamped_add<uint64_t>(tuple_index, frame.end.offset + 1, partition.size());
+    return {start, end};
+  }
+};
+
+template <>
+struct WindowBoundCalculator<FrameType::Range, NoOrderByColumn> {
+  static QueryRange calculate_window_bounds(std::span<const RelevantRowInformation> partition,
+                                            [[maybe_unused]] uint64_t tuple_index,
+                                            [[maybe_unused]] const FrameDescription& frame) {
+    return {0, partition.size()};
+  }
+};
+
+template <typename OrderByColumnType>
+  requires std::is_integral_v<OrderByColumnType>
+struct WindowBoundCalculator<FrameType::Range, OrderByColumnType> {
+  static QueryRange calculate_window_bounds(std::span<const RelevantRowInformation> partition, uint64_t tuple_index,
```

Comment:
Because in Range mode, there can be at most one order-by column and we are in the case where there actually is one (the 0 columns case has `OrderByColumnType = NoOrderByColumn`)